teps/0055-migrate-podlatency-to-taskprelaunchlatency.md

```diff
@@ -0,0 +1,121 @@
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Alternatives](#alternatives)
+  - [Keeping the metrics <code>podLatency</code>](#keeping-the-metrics-)
+  - [Using a map to store the pod <code>startRunning</code> time](#using-a-map-to-store-the-pod--time)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to discard the metrics `podLatency` and add another metrics to
+report the pod creation -> pod running latency.
+
+## Motivation
+
+Currently, `podLatency` calculates the duration between pod creation time and
+pod scheduled time. The `podLatency` value is 0 at 99% of the time because
+Kubernetes timestamps are only at the resolution of a second, so if something
+takes less than a second it looks like it took 0 seconds. At the rest 1% of
+time, the metrics value is 1 or 2, which is not very worth to report. Therefore,
+this TEP proposes to discard the metrics `podLatency` and add another metrics to
+report the pod creation -> pod running latency. It can indicate how long a
+TaskRun pod takes to get ready for running.
+
+### Goals
+
+- Discard the current metrics `podLatency`
+- Add a metrics called `taskPrelaunchLatency` to indicate the pod creation ->
+  pod running latency
+
+### Non-Goals
+
+- Add multiple metrics to indicate different phases of TaskRun initialization,
+  like TaskRun creation time -> pod scheduled time latency, pod scheduled time
+  -> pod running time, etc
+
+## Proposal
+
+As `podLatency` is discarded, `taskPrelaunchLatency` is introduced to report the
+duration between pod creation time, and the time when pod phase turns to
+`Running`. However, Knative doesn't collect and report when a pod starts
+running. The time can be collected in the reconciliation loop. When the first
+time we find the pod phase turns to `Running`, the current timestamp should be
+saved in TaskRun status as `RunAt` time.
+
+```go
+type TaskRunStatusFields struct {
+  // PodName is the name of the pod responsible for executing this task's steps.
+  PodName string `json:"podName"`
+
+  // StartTime is the time the build is actually started.
+  // +optional
+  StartTime *metav1.Time `json:"startTime,omitempty"`
+
+  // CompletionTime is the time the build completed.
+  // +optional
+  CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+
+  // RunAt is the time the taskrun status turns to running
+  // +optional
+  RunAt *metav1.Time `json:"runAt,omitempty"`
+}
+```
+
+Due to the latency of reconciler, the minimum value of `podLatency` becomes 2s.
+
+Compared to `podLatency`, `taskPrelaunchLatency` can report the duration that
+a pod takes to get the primary containers ready and started more accurately. For
+example, a pod might need to download a large image before running. The old
+`podLatency` is 0 in this situation. But `taskPrelaunchLatency` will report the
+downloading time with a 2-second margin.
```

Comment:
I'm afraid I don't understand the 2s margin here.

The taskrun controller will be notified that the taskrun changed, run its reconcile cycle at some point (depending on queue status etc), it will fetch the schedule time and container start time from the pod spec and emit the metric, right? Where does the 2s margin come from?

Reply:
the margin only happened when I tried to get the pod `running` time as I mentioned above. I check the pod status in  reconcile cycle and record the first time when pod status is `running`. It's a stupid way to collect the metric. I think we're fine now.